### ISSUE TITLE
Child Components Not Passed to React

### ISSUE THREAD

**Issue by jjorissen52** [newying61/web-components-with-react#15]
It looks like this does not have child elements being passed to react. My implementation has the exact same problem. Any ideas of how to address that?

**Comment by newying61** [comment]
Hi @jjorissen52 , do you mean pass the light DOM elements into React?
so, something like:
```
// Define the custom element
import { ReactCustomElement } from 'web-components-with-react';
import App from './App';

customElements.define('test-app', ReactCustomElement(App));
```

then in HTML file, render the custom element `test-app` with elements inside it.
```
<test-app>
  <span>An element in light DOM</span>
</test-app>
```

As the `<span>` in the above code is an element in light DOM, so, React App doesn't know what it is.
What you can do is following the custom element specification, add a `<slot></slot>` to your React App.
```
const App = () => {
  return (
    <div>React App</div>
    <slot></slot>
  );
}
```
Then the light DOM elements will be displayed in the slot position.

If you don't want to use a slot but try to put the light DOM elements anywhere, the solution maybe find the shadowRoot, then use `.host` to find the custom element. Then use `.innerHTML` to get the light DOM as a string. But if the light DOM changes, the React App would miss it. Depends on your use case.
this approach needs some DOM query and manipulation.

----------
Just saw the code in your repo, hope I understand it correctly.
It seems you are rendering a `slot` as a child of the React component, then the React component must render the children.
something like this:
```
// In your mRender function:
return createElement(ReactComponent, this.props, createElement('slot')) // slot is passed in as a child
```
The ReactComponent should be something like this:
```
const Button = ({ children }) => {
  // Render the react component and the children (which is the slot you passed in)
  return (
    <button>
      <span>The button text</span>
      {children}
    </button>
  );
}
```
Have a try and see whether it's working :)

**Comment by jjorissen52** [comment]
Very helpful! I read that this is what slots are for but I wasn't able to connect the dots before. I will give this a try.

**Comment by jjorissen52** [comment]
Shadow DOM doesn't help in my case because I need external styles to be applied.

**Comment by newying61** [comment]
Yeah, if you use shadow DOM, then you need to add the styles into it.
this lib has a parameter called `useShadowDOM` (default true) if you turn it off, then the React App will be rendered in light DOM.

By the way, to add the style into shadow DOM, there are several ways.
1 in Webpack, the html plugin has a mount point for the style (don't remember clearly...)
2 use dom method to load the styles into shadow DOM. 
I created a function loadCssFile (https://www.npmjs.com/package/common-dom-utils#load-css-file-into-dom-container)
to do that before. :)
```
 loadCssFile('https://css-file-url/file.css', shadowRoot);
```

But if you don't need shadow DOM, then all good. Easier. :)

**Comment by jjorissen52** [comment]
I was originally using the shadow DOM, but I could not figure out how to make it work with the way Material UI does styling. Okay, so I tried what you suggested, but the React component still appears to be rendering an empty slot. I am wondering how I do it with this library?

```javascript
import * as React from 'react'
import { ReactCustomElementWithProps } from 'web-components-with-react';

const AnotherButton = ({ children }) => <button>
    Stuff should be there => {children}
    <slot></slot>
</button>

customElements.define('another-button', ReactCustomElementWithProps(AnotherButton, ['children'], false))
```

```html
<div class="container">
    <div class="row justify-content-center">
        <div class="col-md-8">
            <div class="card">
                <div class="card-header">Dashboard</div>

                <div class="card-body" id="mbody">
                    <another-button>hello?</another-button>
                </div>
            </div>
        </div>
    </div>
</div>
```

![image](https://user-images.githubusercontent.com/12461302/128579002-2bb331a3-f6e4-4ee6-adea-111d339060fe.png)

**Comment by jjorissen52** [comment]
Looks like it might not be supported (by Web Components) unless you are using the shadow DOM :(

**Comment by newying61** [comment]
When using this lib rendering React component into light DOM, the render function is replacing the light DOM elements with the React app... I think that's the default behaviour of the render function in react-dom. From https://reactjs.org/docs/react-dom.html#render
```
ReactDOM.render() controls the contents of the container node you pass in.
Any existing DOM elements inside are replaced when first called.
Later calls use React’s DOM diffing algorithm for efficient updates.
```
To make it work, I think you need to inherit the `ReactCustomElementWithProps` and in it's renderApp function do something like
```
        // save light DOM elements
        let lightElements = null;
        if (!useShadowDOM) {
          lightElements = this.renderRoot.innerHTML;
        }

      ReactDOM.render(
        <Component renderRoot={this.renderRoot} {...this.props} lightElements={lightElements} />,
        this.renderRoot as Element
      );
```
then pass the innerHTML string to the React App and use `dangerouslySetInnerHTML` to add it in. (remember to sanitise it, use something like DOMPurify)

Or do some DOM manipulation like detach the light DOM node then attach it back to the place you want.

**Comment by newying61** [comment]
For material-ui, it seems there is a `StyleProvider` which can config where to insert the styles.

Please refer the second answer below:
https://stackoverflow.com/questions/51832583/react-components-material-ui-theme-not-scoped-locally-to-shadow-dom
